Feature: better UnitScalar and UnitArray str representation

jonathanrocher opened the pull request:
This implements `__str__` on `UnitScalar` and `UnitArray` classes. 

It allows the output of a `print` for an instance to look nicer, in particular by displaying the unit's label, for example:
`UnitScalar (kilometer): 1` instead of `UnitScalar(1, units='1000.0*m')`

Also, assuming that a new unit is defined:

```
>>> new_unit = copy(dimensionless)
>>> new_unit.label = "Cool unit"
```

printing a UnitScalar or UnitArray will lead to `UnitScalar (Cool unit): 1` instead of `UnitScalar(1, units='1')` or `UnitArray (Cool unit): [1, 2, 3]` instead of `UnitArray([1, 2, 3], units='1')`.

A by-product of this PR is also for the `repr` output to use `.format` instead of the old style `%` formatting and for it to behave better under subclassing (uses `type(self).__name__` instead of writing `UnitScalar` directly).

jonathanrocher commented:
@timdiller Any feedback on this?

timdiller commented on scimath/units/unit_array.py:
```diff
@@ -68,17 +68,13 @@ class UnitArray(numpy.ndarray):
 
+    def __str__(self):
+        base_str = self._get_values_base_str()
+        return "UnitArray(%s, units='%s')" % (base_str, self.units.label)
```
I'm partial to `str.format` myself, but I won't insist... I do prefer to see the use of `self.__class__.__name__`, which behaves better under subclassing:
```
return "{}({}, units={!r})".format(self.__class__.__name__, base_str, self.units.label)
```

rkern replied:
Bikeshedding, but I like this style:

```
return '{0.__name__}({1}, units={2.units.label!r})'.format(type(self), base_str, self)
```

jonathanrocher replied:
Thanks for the comments. Agree with all of them. I find @timdiller 's version a little easier to read, so I used mostly his but with `type(self)`.

rkern commented on scimath/units/unit_scalar.py:
```diff
@@ -17,5 +17,9 @@ class UnitScalar(UnitArray):
+
+    def __str__(self):
+        return ("UnitScalar(%s, units='%s')"
+                % (self.item().__str__(), self.units.label))
```
Just don't implement this function. It looks like an `eval()`able `__repr__` string but isn't.

jonathanrocher replied:
Thanks for the comment @rkern but I don't understand it:
1. I am doing here what I was doing for arrays
2. The output for something like `UnitScalar(1, units="cm")` can be `eval()`-ed because "cm" is parsed. If people create their own units, and add them to their parser, it will be parsed.
3. I always thought that `str` wasn't required to return something that could be `eval()`-ed and was just about returning something readable.

Sorry if I am missing something obvious.

rkern replied:
I wouldn't try to make it *look* evalable when it's not. E.g. just `'%s %s'`

jonathanrocher commented:
I have taken into account feedback from @rkern @timdiller and @cfarrow . I have also added tests and updated the description of the PR. Ready for another look...

rkern commented on scimath/units/unit_array.py:
```diff
@@ -68,16 +68,22 @@ class UnitArray(numpy.ndarray):
-            base_str = base_str[start+1:end]
+        """ String representation using the repr of the unit."""
+        base_str = self._get_values_base_str()
+        s = "{klass}({val}, units='{unit}')"
```
```
s = "..., units={unit!r}"
...
return s.format(..., unit=self.units)
```

senganal commented on scimath/units/unit_array.py:
```diff
@@ -471,6 +477,22 @@ def as_units(self, new_units):
+            base_str = base_str[start+1:end]
+
+        return base_str
+
```
nit: I think a regex match makes the intent clearer.. For e.g.
```
VALUE_PATTERN = re.compile('.*(?P<value>\[.*\]).*')
...
base_str ="{!r}".format(self.view(numpy.ndarray))
matches = VALUE_PATTERN.match(base_str)
if matches:
    base_str = base_str['value']
return base_str
```

stefanoborini replied:
I also suggest the regexp solution, it's a bit more robust and clear.

mdickinson replied:
> it's a bit more robust and clear

I'd have to disagree with this: it's clear that the non-regexp solution is finding the first opening parenthesis and the last closing parenthesis - i.e., that it's finding a pair of matching parentheses. With the regex solution (presuming that @senganal meant something like `re.compile(r'.*(?P<value>\(.*\)).*')`), it's not clear at a glance that we're finding a matching pair of parentheses at all, and if for whatever reason the representation of the array contains inner parentheses, we'll end up with the wrong thing.

```
>>> import numpy as np
>>> dt = np.dtype([('x', float), ('y', float)])
>>> x = np.zeros(2, dtype=dt)
>>> repr(x)
"array([(0.0, 0.0), (0.0, 0.0)], \n      dtype=[('x', '<f8'), ('y', '<f8')])"
>>> import re
>>> re.compile(r'.*(?P<value>\(.*\)).*').match(repr(x)).group('value')
'(0.0, 0.0)'
```

stefanoborini replied:
Right. good as is then.

senganal commented:
@jonathanrocher  Did a master merge and added one minor comment..

senganal commented:
also, please update `CHANGES.txt`

jonathanrocher commented:
Tried to revive this old PR. Added an entry in the changelog as asked by @senganal and resolved conflicts with master. Ok to merge?
